secretstores/hashicorp/vault/vault.go

```diff
@@ -197,55 +236,96 @@ func (v *vaultSecretStore) getSecret(secret string) (*vaultKVResponse, error) {
 
 	var d vaultKVResponse
 
-	if err := json.NewDecoder(httpresp.Body).Decode(&d); err != nil {
-		return nil, fmt.Errorf("couldn't decode response body: %s", err)
+	if v.vaultValueType.isMapType() {
+		// parse the secret value to map[string]string
+		if err := json.NewDecoder(httpresp.Body).Decode(&d); err != nil {
+			return nil, fmt.Errorf("couldn't decode response body: %s", err)
+		}
+	} else {
+		// treat the secret as string
+		b, err := ioutil.ReadAll(httpresp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't read response: %s", err)
+		}
+		res := v.json.Get(b, DataStr, DataStr).ToString()
+		d.Data.Data = map[string]string{
+			secret: res,
+		}
 	}
 
 	return &d, nil
 }
 
 // GetSecret retrieves a secret using a key and returns a map of decrypted string/string values
 func (v *vaultSecretStore) GetSecret(req secretstores.GetSecretRequest) (secretstores.GetSecretResponse, error) {
-	d, err := v.getSecret(req.Name)
+	// version 0 represent for latest version
+	version := "0"
+	if value, ok := req.Metadata[versionID]; ok {
+		version = value
+	}
+	d, err := v.getSecret(req.Name, version)
 	if err != nil {
 		return secretstores.GetSecretResponse{Data: nil}, err
 	}
 
 	resp := secretstores.GetSecretResponse{
-		Data: map[string]string{},
-	}
-
-	// Only using secret data and ignore metadata
-	// TODO: add support for metadata response when secretstores support it.
-	for k, v := range d.Data.Data {
-		resp.Data[k] = v
+		Data: d.Data.Data,
 	}
 
 	return resp, nil
 }
 
 // BulkGetSecret retrieves all secrets in the store and returns a map of decrypted string/string values
 func (v *vaultSecretStore) BulkGetSecret(req secretstores.BulkGetSecretRequest) (secretstores.BulkGetSecretResponse, error) {
+	version := "0"
+
+	resp := secretstores.BulkGetSecretResponse{
+		Data: map[string]map[string]string{},
+	}
+
+	keys, err := v.listKeysUnderPath("")
+	if err != nil {
+		return secretstores.BulkGetSecretResponse{}, err
+	}
+
+	for _, key := range keys {
+		keyValues := map[string]string{}
+		secrets, err := v.getSecret(key, version)
```

Comment:
Can version as be passed in the Bulk Get API? it seems to always default to 0.

Reply:
The reason why I didn't add the version in the Bulk Get API is because that, for one secret if the version is specified but does not exist, none secret information will be returned. Therefore, when obtaining secrets in batches, the version is hard coded to 0 which means the latest version.

If it is better to do like following:
Using version passed in the Bulk Get API, those secrets ​​that do not have the specified version will not appear. If the version is not specified it will defaults to 0.  And write this into the document, to let users know how to use it better.